Mathlib/CategoryTheory/Sites/SheafHom.lean

```diff
@@ -0,0 +1,185 @@
+        rintro ⟨X₁⟩ ⟨X₂⟩ ⟨f : X₂ ⟶ X₁⟩
+        dsimp
+        refine' Eq.trans _ ((s.1 ⟨X₁⟩).naturality
+          (Over.homMk f : Over.mk f ⟶ Over.mk (𝟙 X₁)).op)
+        erw [← s.2 f.op, internalHom_map_app_op_mk_id]
+        rfl }
+  invFun f := ⟨fun X => whiskerLeft _ f, fun _ => rfl⟩
+  left_inv s := by
+    dsimp
+    ext ⟨X⟩ ⟨Y : Over X⟩
+    have H := s.2 Y.hom.op
+    dsimp at H ⊢
+    rw [← H]
+    apply internalHom_map_app_op_mk_id
+  right_inv f := rfl
+
+variable {F G}
+
+lemma InternalHom.isAmalgamation_iff {X : C} (S : Sieve X)
+    (x : Presieve.FamilyOfElements (internalHom F G) S.arrows)
+    (hx : x.Compatible) (y : (internalHom F G).obj (op X)) :
+    x.IsAmalgamation y ↔ ∀ (Y : C) (g : Y ⟶ X) (hg : S g),
+      y.app (op (Over.mk g)) = (x g hg).app (op (Over.mk (𝟙 Y))) := by
+  constructor
+  · intro h Y g hg
+    rw [← h g hg, internalHom_map_app_op_mk_id]
+  · intro h Y g hg
+    dsimp
+    ext ⟨W : Over Y⟩
+    refine (h W.left (W.hom ≫ g) (S.downward_closed hg _)).trans ?_
+    have H := hx (𝟙 _) W.hom (S.downward_closed hg W.hom) hg (by simp)
+    dsimp at H
+    simp only [Functor.map_id, FunctorToTypes.map_id_apply] at H
+    rw [H, internalHom_map_app_op_mk_id]
+    rfl
+
+variable (F G)
+
+lemma internalHom_isSheafFor {X : C} (S : Sieve X)
+    (hG : ∀ ⦃Y : C⦄ (f : Y ⟶ X), IsLimit (G.mapCone (S.pullback f).arrows.cocone.op)) :
+    Presieve.IsSheafFor (internalHom F G) S.arrows := by
+  intro x hx
+  apply exists_unique_of_exists_of_unique
+  · have Φ : ∀ {Y : C} (g : Y ⟶ X), ∃ (φ : F.obj (op Y) ⟶ G.obj (op Y)),
+      ∀ {Z : C} (p : Z ⟶ Y) (hp : S (p ≫ g)), φ ≫ G.map p.op =
+        F.map p.op ≫ (x (p ≫ g) hp).app ⟨Over.mk (𝟙 Z)⟩ := fun {Y} g => by
+        let c : Cone ((Presieve.diagram (Sieve.pullback g S).arrows).op ⋙ G) :=
+          { pt := F.obj (op Y)
+            π :=
+              { app := fun ⟨Z, hZ⟩ => F.map Z.hom.op ≫ (x _ hZ).app (op (Over.mk (𝟙 _)))
+                naturality := by
+                  rintro ⟨Z₁, hZ₁⟩ ⟨Z₂, hZ₂⟩ ⟨f : Z₂ ⟶ Z₁⟩
+                  dsimp
+                  rw [id_comp, assoc]
+                  have H := hx f.left (𝟙 _) hZ₁ hZ₂ (by simp)
+                  simp only [internalHom_obj, unop_op, Functor.id_obj, op_id,
+                    FunctorToTypes.map_id_apply] at H
+                  let φ : Over.mk f.left ⟶ Over.mk (𝟙 Z₁.left) := Over.homMk f.left
+                  have H' := (x (Z₁.hom ≫ g) hZ₁).naturality φ.op
+                  dsimp at H H' ⊢
+                  erw [← H, ← H', internalHom_map_app_op_mk_id, ← F.map_comp_assoc,
+                    ← op_comp, Over.w f] } }
+        use (hG g).lift c
+        intro Z p hp
+        exact ((hG g).fac c ⟨Over.mk p, hp⟩)
+    let app : ∀ {Y : C} (_ : Y ⟶ X), F.obj (op Y) ⟶ G.obj (op Y) := fun g => (Φ g).choose
+    have happ : ∀ {Y : C} (g : Y ⟶ X) {Z : C} (p : Z ⟶ Y) (hp : S (p ≫ g)),
+      app g ≫ G.map p.op = F.map p.op ≫ (x (p ≫ g) hp).app ⟨Over.mk (𝟙 Z)⟩ :=
+        fun g => (Φ g).choose_spec
```

Comment:
Or instead of the `have` statement above, just define these directly without `Exists.choose` and `Exists.choose_spec`

Reply:
Thanks for the suggestions! I have made the `have` a separate lemma, and this `app` a separate definition, which splits the construction/proof is smaller chunks.